### DISCUSSION TITLE
How to only select regular list item and not task list item?

### DISCUSSION THREAD

**Discussion by ooker777** [syntax-tree/unist#132] [Q&A]
To select only `listItem`, I use:
```js
import {fromMarkdown} from 'mdast-util-from-markdown'
import {inspect} from 'unist-util-inspect'
import {selectAll} from 'unist-util-select'
import {gfm} from 'micromark-extension-gfm'
import {gfmFromMarkdown, gfmToMarkdown} from 'mdast-util-gfm'

const doc = `
## List 
- One
- Two 
## Tasklist
* [ ] to do
* [x] done
`
const tree = fromMarkdown(doc, {
  extensions: [gfm()],
  mdastExtensions: [gfmFromMarkdown()]
})

const listItems = selectAll("listItem", tree)
console.log(inspect(listItems))
```

The output:
```
├─0 listItem[1] (3:1-3:6, 10-15)
│   │ spread: false
│   │ checked: null
│   └─0 paragraph[1] (3:3-3:6, 12-15)
│       └─0 text "One" (3:3-3:6, 12-15)
├─1 listItem[1] (4:1-4:7, 16-22)
│   │ spread: false
│   │ checked: null
│   └─0 paragraph[1] (4:3-4:7, 18-22)
│       └─0 text "Two" (4:3-4:6, 18-21)
├─2 listItem[1] (6:1-6:12, 35-46)
│   │ spread: false
│   │ checked: false
│   └─0 paragraph[1] (6:7-6:12, 41-46)
│       └─0 text "to do" (6:7-6:12, 41-46)
└─3 listItem[1] (7:1-7:11, 47-57)
    │ spread: false
    │ checked: true
    └─0 paragraph[1] (7:7-7:11, 53-57)
        └─0 text "done" (7:7-7:11, 53-57)
```
I notice that task list items have boolean `checked`, while regular list items have null `checked`. So to select only task list items, I can use:
```js
const taskItems = selectAll("[checked]", tree)
```
But how can I only select regular list item? Either of these doesn't work:
```js
const regularItems = selectAll("[checked=null]", tree)  // [] 
const regularItems = selectAll(":not([checked])", tree) // select all nodes
const regularItems = filter(listItems, node => node.checked === null) // undefined
const regularItems = filter(listItems, node => node.checked !== true && node.checked !== false) //undefined
```
I know I can work on the text, but that's cumbersome and defeats the purpose of the AST.

([Also asked on Stack Overflow](https://stackoverflow.com/q/78913847/3416774))

**Comment by ChristianMurphy** [comment]
Hey again @ooker777! 👋 
What you seem to want is tree traversal https://unifiedjs.com/learn/recipe/tree-traversal/
You could also use select or filter, but a visitor pattern will probably be the easiest.
This is also a place where a unified pipeline could simplify the structure a bit.

<details>

<summary>example of a pipeline that gets list items</summary>

```ts main.ts
import type {Root} from "mdast"
import {remark} from 'remark'
import remarkGfm from 'remark-gfm'
import {visit} from 'unist-util-visit'

const sourceMarkdown = `
## List 
- One
- Two
## Tasklist
* [ ] to do
* [x] done
`

function remarkListPlugin () {
  return (tree: Root) => {
    visit(tree, "listItem", (node) => {
      if (node.checked === null) {
        // Do something with the list item nodes
        console.log(node)
      }
    })
  }
}

try {
  // this renders the source to the page
  document.querySelector('#source')!.textContent = sourceMarkdown

  // this is the transformation pipeline
  const file = await remark()
    .use(remarkGfm)
    .use(remarkListPlugin)
    .process(sourceMarkdown)

  // these render the output to the page
  document.querySelector('#result')!.textContent = String(file)
  document.querySelector('#error')!.textContent = ''
} catch (error) {
  document.querySelector('#error')!.textContent = String(error)
}
```

</details>

runnable example: https://codesandbox.io/p/devbox/remark-with-vite-rh4w48

**Reply by ooker777** [reply]
for the sake of learning, how to use it with `filter` or `selectAll`? Also, as I don't need to stringify the tree but just inspect it, why do you see that using the transform pipeline/process is easier? I can see that if I decide to stringify it later, then I have to rewrite the code, which isn't ideal. But let's say I'm sure that it won't happen, then I see that this approach require you to write more boilerplate, no?

**Reply by ChristianMurphy** [reply]
> Also, as I don't need to stringify the tree but just inspect it

This is likely an [XY Problem](https://xyproblem.info/).
I doubt you _just_ want to inspect it, you want to _do_ something with it, knowing what that something is would help give a better recommendation.

> why do you see that using the transform pipeline/process is easier? 

Without a pipeline you end up writing a bunch of boilerplate code

> then I see that this approach require you to write more boilerplate, no?

I think the reverse is often true

> for the sake of learning, how to use it with filter or selectAll? 

```ts
function remarkListPlugin() {
  return (tree: Root) => {
    const listItems = selectAll("listItem", tree);
    listItems
      .filter((node) => "checked" in node && node.checked === null)
      .forEach((node) => console.log(node));
  };
}
```

**Reply by ooker777** [reply]
Thanks. As for that something, I have a bunch of markdown notes having the same structure:
```
Field 1: value 1
Field 2: value 2
...
## List 
- One
- Two 
- ...
## Tasklist
* [ ] to do
* [x] done
* [ ] ...
```
I need to get the data on them and have a summary table. So I see that I only need to use the transformer to get the data. I don't see the need to stringfiy them. What do you think?

```js
listItems
      .filter((node) => "checked" in node && node.checked === null)
```
Oh, you are using the regular [`filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter "Array.prototype.filter() - JavaScript | MDN") method for arrays? I was looking to learn more about https://github.com/syntax-tree/unist-util-select and https://github.com/syntax-tree/unist-util-filter.

**Reply by ChristianMurphy** [reply]
> I was looking to learn more about https://github.com/syntax-tree/unist-util-select and https://github.com/syntax-tree/unist-util-filter.

These are two different ways to get at nodes.
My recommendation is use one of the other, don't mix them, mixing will complicate things.

If you do mix them as a learning exercise, then treat it as a learning exercise. 🙂 
Take some time to understand them, and try out different things, experiment.

The goal

> I need to get the data on them and have a summary table

In HTML, in React, in JSON?
What kind of a summary?

> I don't see the need to stringify them. What do you think?

The compiler step most often stringifies, but it doesn't have to.
React markdown for example returns react components instead.

**Reply by ooker777** [reply]
No I don't intend to mix them. I just need to use one of them. I do want to challenge myself a bit, but I don't intend to do the hardest way. My goal now is to have a good mental model of the user interface, not at the advanced level.

Hmm, the final output is HTML, but I have another lib to handle that, so my output on this task is just an object I'd say. 

Hmm, so when you say using the process is cleaner, you mean that it is always cleaner when you consider the final output?

**Reply by wooorm** [reply]
> Hmm, the final output is HTML, but I have another lib to handle that, so my output on this task is just an object I'd say.

Can you please expand? On the lib. On what your actual goal is?

> Hmm, so when you say using the process is cleaner, you mean that it is always cleaner when you consider the final output?

It depends, on your goals

**Reply by ooker777** [reply]
So I'm using Obsidian with the dataview plugin. The plugin can make the summary table but it's limited and rigid in the format. So I make my own plugin to parse the notes and use the dataview api to render the table. Is it clearer?

**Reply by wooorm** [reply]
Are you creating an obsidian plugin?
What is a summary table? What output does the plugin currently generate, from what input, and what is the output you instead want?

**Reply by ooker777** [reply]
yes you can say I'm creating an Obsidian plugin. A summary table is just a table that gather information from various sources. It's just a term I use for my task, not a jargon in Obsidian or the Dataview plugin. I will use the [`dv.table(headers, elements)`](https://blacksmithgu.github.io/obsidian-dataview/api/code-reference/#dvtableheaders-elements "Codeblock Reference - Dataview") function to render a table. 

# Input 
Many files with this template:
```js
Field 1: value 1
Field 2: value 2
...
## List 
- One
- Two 
- ...
## Tasklist
* [ ] to do
* [x] done
* [ ] ...
```
# Output
```js
const elements = [
  ["value 1", "value 2", ["One", "Two"], ["todo", "done"]],
  /// ...
]

**Reply by wooorm** [reply]
- There seems to be some frontmatter-like format? How does this custom syntax work? When does it stop? Is it YAML but without a fence? What is it? Why are the keys dropped?
- Your examples of list items are each a word. What if there’s other things in list items. Emphasis. Code blocks. etc?
- What is the type of `elements`? `Array<Array<string> | string>`? Why is there only one item. How do you get more items?

**Reply by ooker777** [reply]
- No it's not YAML without a fence. The whole input is pure markdown. (Technically there is a frontmatter, but the input is outside of it). I don't get "Why are the keys dropped?" means.
- I can be sure that only plain text be used
- Yes, the type of `elements` is indeed `Array<string | string[]>`. Each file in the folder will make an element. I will read all the files in the folder, extract the data and push it into `elements`.

**Reply by wooorm** [reply]
* I don’t understand how for markdown `Field 1: value 1`, you suddenly get `'value 1'` in a javascript array
* If you can be so sure (which I wonder about? People always like adding bold and such), why use markdown? You can use a structured data format: YAML or so.
* why are some things `string` (`value 1`) and some things `Array<string>`?

**Reply by ooker777** [reply]
- I use this code to extract `value 1`:
```js
const text = select("paragraph text", sectionNode)!.value;
  const fields = text.split("\n");
  for (const field of fields){
    const [key, value] = field.split(":")
    data[key] = value
  }
```
- Tables and tasklists are structured data that YAML doesn't support. In Obsidian the frontmatter should be used for plugin settings and not user data, and it will be hidden depending on user preference. I suppose I can make change the editor behavior on those data pages, but I guess there are unstructured notes that user would want to insert on their own. I'll think about this more, thanks. 
- Because the data is so? I don't know how to answer this question better

**Reply by wooorm** [reply]
* I don’t mean the code you currently drafted. I mean **why** do you want this? Why did you invent this data format? And, the code you now show is different than what you showed before. The code `data[key] = value`, for the input `Field 1: value 1`, cannot output `["value 1", …`. For me to help you well, I really need to know what you want and why.
* Arrays of arrays are tables, and booleans can be used for tasklists; why do you use obsidian/markdown if you don‘t allow anything from markdown other than paragraphs and lists? I was under the impression that you were the only user, as you know for certain only paragraphs/lists are used, but this point seems to indicate there are many unknown users?
* I don’t understand how `Field 1: value 1` turns into `value 1`. Why drop `Field 1`? Why store this as an array, where it seems like an object?

---

Could you perhaps rephrase your example from https://github.com/orgs/syntax-tree/discussions/132#discussioncomment-10476919 again? Who are the people writing that? Do you have an actual real non-abbreviated example of what they write? And can you then, by hand, make the exact non-abbreviated output you want to generate? I can help you better if I know what you want.

**Reply by ooker777** [reply]
> why do you want this? Why did you invent this data format? 

It's just to have most convenience to write the input. Perhaps there is a better, more standard way but I don't know.
> The code `data[key] = value`, for the `input Field 1: value 1`, cannot output `["value 1", …`

The data of a note will be fill into the `data` object. Then it will be pushed/flatten to the `elements` array:
```js
elements.push([data["Field 1"], data["Field 2"], ..., [data["List"][0], data["List"][1], ...], [data["Tasklist"][0], data["Tasklist"][1], ...]])
```
> I don’t understand how `Field 1: value 1` turns into `value 1`. Why drop `Field 1`? Why store this as an array, where it seems like an object?

It is used in the headers of the table, which is stored in a separate variable. I store this as an array because that how the API I use to render the table requires ([`dv.table(headers, elements)`](https://blacksmithgu.github.io/obsidian-dataview/api/code-reference/#dvtableheaders-elements "Codeblock Reference - Dataview")). But yeah I'm considering not using it anymore and just build the table with HTML.

> why do you use obsidian/markdown if you don‘t allow anything from markdown other than paragraphs and lists? 

I allow anything, but only need data from tables, lists, tasklists, headings, paragraphs with colons. Other else can be ignored.
> I was under the impression that you were the only user, as you know for certain only paragraphs/lists are used, but this point seems to indicate there are many unknown users?

I write it for my needs first, but I will also publish it for other people to use. If they use my plugin they will need to follow my structure I guess?

> Do you have an actual real non-abbreviated example of what they write? And can you then, by hand, make the exact non-abbreviated output you want to generate?

Is this what you are looking for? It's all the data for a coffee shop. The header cells of the table are their shop locations. The lists are for their strengths and weaknesses.
```md
Giá nước: 0
Giờ mở cửa: 8:00 - 21:00
Liên hệ: 867990125
Website: [The best cafe](https://tonkin.coffee/aboutbestcoffee/)
Facebook: [Tonkin Specialty Coffee | Ho Chi Minh City | Facebook](https://www.facebook.com/tonkinspecialtycoffee)

| Chi nhánh  | 91 Lý Tự Trọng, Q1                           | 135/50 Trần Hưng Đạo, Q1 |
| ---------- | -------------------------------------------- | ------------------------ |
| Giá giữ xe | 10k                                          | 0                        |
| Số lượng   | 30 người                                     | 10 người                 |
| Ghi chú    | Lối vào nhỏ, dễ khuất, phải nhìn kỹ mới thấy |                          |
| Lưu ý      | Xe lăn không tiếp cận được                   |                          |
| Điểm cộng  | Không gian ngoài trời                        |                          |

## Lưu ý
- Không có loa và mic
- Cần đặt chỗ trước

## Ghi chú
- Chỉ miễn phí với các sự kiện phi lợi nhuận
```

**Reply by wooorm** [reply]
Yes. That is exactly what I am looking for. Thank you!
Can you also post the exact output you want? Can you, by hand, copy/paste the parts of that markdown, and make it into a table? How do you want that table to look?

**Reply by ooker777** [reply]
```js
const data = {
  Name: "Tonkin",
  "Location type": "Quán có phòng riêng",
  "Price": 0,
  "Opening hour": [new Date("8:00"), new Date("21:00")],
  "Contact": "867990125", 
  Website: 'https://tonkin.coffee/aboutbestcoffee/',
  Facebook: "https://www.facebook.com/tonkinspecialtycoffee",
  "Locations": [
    {
      "Address": "91 Lý Tự Trọng, Q1",
      "Parking price": 10000,
      "Quantity": 30,
      "Notes": ["Lối vào nhỏ, dễ khuất, phải nhìn kỹ mới thấy"],
      "Keep in mind": ["Xe lăn không tiếp cận được"],
      "Bonus": ["Không gian ngoài trời"],
    },
    {
      "Address": "135/50 Trần Hưng Đạo, Q1",
      "Parking price": 0,
      "Quantity": 10,
    }
  ],
  "Keep in mind": ["Không có loa và mic", "Cần đặt chỗ trước"],
  "Notes": ["Chỉ miễn phí với các sự kiện phi lợi nhuận"],
}
const headers = ["Name", "Quantity", "Price", "Opening hour", "Keep in mind", "Address", "Bonus" ]
const elements = [
  [
    "Tonkin 1",
    30,
    0,
    [new Date("8:00"), new Date("21:00")],
    ["Xe lăn không tiếp cận được", "Không có loa và mic", "Cần đặt chỗ trước"],
    "91 Lý Tự Trọng, Q1",
    ["Không gian ngoài trời"]  
  ], 
  [
    "Tonkin 2",
    30,
    0,
    [new Date("8:00"), new Date("21:00")],
    ["Không có loa và mic", "Cần đặt chỗ trước"],
    "135/50 Trần Hưng Đạo, Q1",
  ], 
] 
```
> copy/paste the parts of that markdown, and make it into a table

I don't get what parts of that markdown it is, but the output table would be:

| Name     | Quantity | Price | Opening hour | Keep in mind                                                                              | Address                  | Bonus                 |
| -------- | -------- | ----- | ------------ | ----------------------------------------------------------------------------------------- | ------------------------ | --------------------- |
| Tonkin 1 | 30       | 0     | 8:00 - 21:00 | <li>Xe lăn không tiếp cận được</li><li>Không có loa và mic</li><li>Cần đặt chỗ trước</li> | 91 Lý Tự Trọng, Q1       | Không gian ngoài trời |
| Tonkin 2 | 10       | 0     | 8:00 - 21:00 | <li>Không có loa và mic</li><li>Cần đặt chỗ trước</li>                                    | 135/50 Trần Hưng Đạo, Q1 |                       |

**Reply by wooorm** [reply]
Sorry, but I am looking for input which maps to output.
You gave me input markdown, about water price for example, which is not present in your output.
And your output includes fields such as `Name` and `Website`, which do not exist in your input markdown.
Also, beware that `new Date("8:00")` is not valid. That is broken. I am not sure what you want with that.
Then, your input is Vietnamese, but in JavaScript it is suddenly in English. Where do you imagine this translation to occur?

I am sorry. It is unclear to me how I can help you best. I do not understand what you want.

**Reply by ooker777** [reply]
Sorry for making the situation more complicated. At the beginning I sent you the Vietnamese markdown input, but after that I think it would be a barrier for you to read my intention. So I translate the keys for you the output. I didn't translate the values, as I think you'd need to compare it to the already given input. In my real case I'll use Vietnamese in both input and output. With you I'll translate both, but if you need the Vietnamese version I'll provide.

Let's just ignore what I send above and use the below instead.
# Input
Filepath: `Coffee shop with meeting room/Tonkin.md`
```md
Price: 0
Opening hour: 8:00 - 21:00
Contact: 867990125
Website: [The best cafe](https://tonkin.coffee/aboutbestcoffee/)
Facebook: [Tonkin Specialty Coffee | Ho Chi Minh City | Facebook](https://www.facebook.com/tonkinspecialtycoffee)

| Locations     | 91 Road A, District 1                                    | 135/50 Street B, District 1 |
| ------------- | -------------------------------------------------------- | --------------------------- |
| Parking price | 10k                                                      | 0                           |
| Quantity      | 30 people                                                | 10 people                   |
| Notes         | Small entry and hidden from view, have to look carefully |                             |
| Keep in mind  | Wheelchair inaccessible                                  |                             |
| Bonus         | Outdoor                                                  |                             |

## Keep in mind
- No speaker and microphone
- Reservation required

## Notes
- Only free for non-profit events
```
# Output
```js
const data = {
  Name: "Tonkin",  // infer from basename
  "Location type": "Coffee shop with meeting room",  // infer from path
  "Price": 0,
  "Opening hour": getOpeningHour("8:00 - 21:00"),  // return's type: [Date, Date] 
  "Contact": "867990125",
  Website: new URL("https://tonkin.coffee/aboutbestcoffee/"),
  Facebook: new URL("https://www.facebook.com/tonkinspecialtycoffee"),
  "Locations": [
    {
      "Address": "91 Road A, District 1",
      "Parking price": 10000,
      "Quantity": 30,
      "Notes": ["Small entry and hidden from view, have to look carefully"],
      "Keep in mind": ["Wheelchair inaccessible"],
      "Bonus": ["Outdoor"],
    },
    {
      "Address": "135/50 Street B, District 1",
      "Parking price": 0,
      "Quantity": 10,
    }
  ],
  "Keep in mind": ["No speaker and microphone", "Reservation required"],
  "Notes": ["Only free for non-profit events"],
} 
// Header values are selected and ordered by users
const headers = ["Name", "Quantity", "Price", "Opening hour", "Keep in mind", "Address", "Bonus"];
const elements = [
  [
    "Tonkin 1",
    30,
    0,
    "8:00 - 21:00",
    ["Wheelchair inaccessible", "No speaker and microphone", "Reservation required"],
    "91 Road A, District 1",
    ["Outdoor"]
  ],
  [
    "Tonkin 2",
    30,
    0,
    "8:00 - 21:00",
    ["No speaker and microphone", "Reservation required"],
    "135/50 Street B, District 1",
  ],
];
```
| Name     | Quantity | Price | Opening hour | Keep in mind                                                                                    | Address             | Bonus   |
| -------- | -------- | ----- | ------------ | ----------------------------------------------------------------------------------------------- | ------------------- | ------- |
| Tonkin 1 | 30       | 0     | 8:00 - 21:00 | <li>Wheelchair inaccessible</li><li>No speaker and microphone</li><li>Reservation required</li> | 91 Road A, Q1       | Outdoor |
| Tonkin 2 | 10       | 0     | 8:00 - 21:00 | <li>No speaker and microphone</li><li>Reservation required</li>                                 | 135/50 Street B, Q1 |         |

I hope this is better. Thank you for your patience.
